include/albatross/src/utils/vector_utils.hpp

```diff
@@ -31,6 +31,30 @@ inline std::vector<double> linspace(double a, double b, std::size_t n) {
   return xs;
 }
 
+inline bool all(const std::vector<bool> &xs) {
```

Comment:
Uh oh, you've stumbled into the dreaded `std::vector<bool>`! There's a specialization in the standard for `std::vector<bool>` which attempts to save space by storing each `bool` in a single bit. The consequence of this is that when you iterate over a `std::vector<bool>` you don't get actual references you get reference like classes.

TL;DR: I think the `for` loops below probably shouldn't make `x` be a reference.